### ISSUE TITLE
Not an issue

### ISSUE THREAD

**Issue by c0debreaker** [jenkins201/packer-terraform-cicd-aws#1]
I really don't want to post it here because it's not an issue about your repo.

It's really about me getting stuck right at the beginning. I'm not sure where to start, how to get it up and running. I do have the cli tools like terraform and packer installed.

I've executed `docker run -it --rm simonmcc/hashicorp-pipeline:latest /bin/ash` however, I don't know where to go next.

**Comment by simonmcc** [comment]
The `packer-terraform-cicd-aws` project is setup to be used inside an existing Jenkins environment, if you don't have an existing Jenkins server, I can recommend using the container I used to work on this project, that's over at https://github.com/jenkins201/jenkins-container, it gives a docker based Jenkins server, that then uses the same docker host to create containers to run the Jenkins jobs in, if you have docker installed on your workstation, should should be able to clone https://github.com/jenkins201/jenkins-container, cd into jerkins-container & then run `docker-compose up`.  This is how I worked on this project, I'm going to try & expand the documentation around this, to make it clearer for people unfamiliar running there own Jenkins server.  Notes on running the Jenkins server are here: https://github.com/jenkins201/jenkins-container/blob/master/README.md

**Comment by c0debreaker** [comment]
I have an existing Jenkins running in a docker container. It's been up and running for a week and have been experimenting on it trying to figure out how to automagically create new jobs using jobdsl or Pipeline with very little luck. :(

Since my Jenkins already exists, how do I use your config?

Or should I just start from beginning and use the link you provided? https://github.com/jenkins201/jenkins-container

It would be nice if I can use my existing so I'll learn more on debugging stuff if something goes wrong.

fyi, I looked at my unix command history and this was the command I ran
```sh
docker run -p 8080:8080 -p 50000:50000 jenkins/jenkins:lts
```

**Comment by simonmcc** [comment]
I you want to use your existing Jenkins setup, you have 2 options:

* Simplest option - you'll need to create a new "Multibranch Pipeline" project, the config of which should look like this:
<img width="977" alt="jenkins201_terraform" src="https://user-images.githubusercontent.com/15053/43266619-c0fc9026-90e3-11e8-959a-6cf42aa7e4dc.png">

* More involved, setup a seed job (just a "Freestyle project") that will use the Job DSL definitions to create new jobs, here's what mine looks like:
<img width="910" alt="jenkins-container-seed-job-1" src="https://user-images.githubusercontent.com/15053/43267841-f18380e4-90e6-11e8-8e21-826014a25e0b.png">
<img width="802" alt="jenkins-container-seed-job-2" src="https://user-images.githubusercontent.com/15053/43267854-f685b378-90e6-11e8-95fe-ce5d9367714c.png">
<img width="822" alt="jenkins-container-seed-job-3" src="https://user-images.githubusercontent.com/15053/43267857-f9c52f50-90e6-11e8-98e1-489852b1f490.png">

(You may need to install some extra Jenkins plugins to get it all to work, here's the list of plugins I install: https://github.com/jenkins201/jenkins-container/blob/master/Dockerfile#L18)

**Comment by c0debreaker** [comment]
Very cool! I'll let you know if I encounter any issue.

**Comment by c0debreaker** [comment]
Hi Simon,

I used your 2nd option. I now have a seed job. Here is a screenshot - https://i.imgur.com/7cubVjE.png

What are the next steps I should do?

Thanks!

**Comment by simonmcc** [comment]
So that single job has generated 3 new jobs for you, based on the `jobs/*.groovy` files in the Jenkins-container repo:

![7cubvje_png_1_704x1_372_pixels](https://user-images.githubusercontent.com/15053/43355283-c28f2aba-9251-11e8-9e09-29a720a74fe8.png)

you'll see those jobs if you click the Jenkins link in the top right hand corner.

The next step would be to run the `jenkins201-packer-terraform-cicd` job, this will probably fail on your first attempt, as you need to create an AWS credential set for it, called `demo-aws-creds` (the name is referenced in the Jenkinsfile: https://github.com/jenkins201/packer-terraform-cicd-aws/blob/master/Jenkinsfile#L43

**Comment by c0debreaker** [comment]
I saw all 3 new jobs. :)

I'll create `demo-aws-creds` in IAM and I will retry `jenkins201-packer-terraform-cicd`.

Let's say I'm going to create a new web application project in Github. Let's assume I finished the web application and it's working. I execute `npm run build` on the command line to generate a production version of this web app. This creates directory named `build`.

What I'd like to achieve is put `npm run build` in `Execute shell` in Jenkins job. However, I'd like a Jenkins job to be created automatically first.

so something like in the groovy
```
steps {
    shell('npm run build')
}
```

**Comment by simonmcc** [comment]
You have 2 options, put it in a Job DSL .groovy file, in which case your example above is right, the other option is to use a Jenkinsfile in your main project repo.

**Comment by c0debreaker** [comment]
When you said `main project repo`, did you mean put `Jenkinsfile` in my new web application repo's root directory?

**Comment by c0debreaker** [comment]
I restarted from scratch and got rid of my own Jenkins server. Earlier I used my own Jenkins server and used your option 2. I was shocked to see my Jenkins server was missing a lot of tools. Yours also had Artifactory. That is very cool!

**Comment by c0debreaker** [comment]
I've been analyzing the `jenkins-container` repo. I think I finally know now how the 3 other jobs are being created. It's because `packer_terraform_cicd_aws.groovy`, `simple_java_maven_app.groovy` and `terraform_iac.groovy` files are included in this repo specifically inside the jobs directory. Then in DSL Scripts in configuration of the seed job, those files matches the  `jobs/*.groovy` pattern.

At my work(not sure how Jenkins/gitlab was implemented), we don't update the seed job to specify a new job that should be created. The creation of the new job is automated. It happens when we add the file `runway/dsl.groovy` to a repo and when we pushed to master branch.

Using your workflow, how can we achieve the same thing?

**Comment by c0debreaker** [comment]
I almost got the Jenkinsfile to work but I had to stop my Github Organization job. I don't like the approach as it scans my entire account for all the repos I have. If I implement this GO at work, and I set the interval to check every 5 mins, it is not good as we have more than 1500 repositories.

The reason why I picked 5 mins is because currently at work, the way new jenkins jobs are created are done via server side git hooks which triggers the jenkins seed job. This is what I was told. The server side hook is triggered when a user git push to his master branch and if `runway/dsl.groovy` file exists.

**Comment by simonmcc** [comment]
The default Github Org plugin scans an entire GitHub org, but you don't have to use that or even do that - my initial Job DSL just used standard git polling against a git repo hosted on GH.  You can also tel the GitHub Org plugin to watch a single repo in an org too.

**Comment by c0debreaker** [comment]
Ah, that's what I didn't know. Thanks for letting me know. I finally was able to reverse engineer our gitlab server git hooks at work. I installed a gitlab on my virtualbox. I still used your Jenkins container. I wrote a server side hook(filename is update) and it reads the dsl.groovy from the push I made. I then make an api call with  dsl script as part of the payload and it submits it to the seed job. The seed job will process the payload. If it's valid, it will generate the job mentioned on the submitted dsl script. It's working great!
